s3.go

```diff
@@ -14,20 +20,184 @@ type S3 struct {
 
 // Open implements FS.
 func (s *S3) Open(ctx context.Context, path string) (*File, error) {
-	return nil, fmt.Errorf("Open not implemented for S3")
+	sh, err := s.s3Client(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	obj, err := sh.GetObjectWithContext(ctx, &s3.GetObjectInput{
+		Bucket: aws.String(s.Bucket),
+		Key:    aws.String(path),
+	})
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == s3.ErrCodeNoSuchKey {
+				return nil, &notExistError{
+					Path: path,
+				}
+			}
+		}
+		return nil, fmt.Errorf("s3: unable to fetch object: %v", err)
+	}
+
+	return &File{
+		ReadCloser: obj.Body,
+		Name:       path,
+		ModTime:    *obj.LastModified,
+		Size:       *obj.ContentLength,
+	}, nil
 }
 
 // Create implements FS.
 func (s *S3) Create(ctx context.Context, path string) (io.WriteCloser, error) {
-	return nil, fmt.Errorf("Create not implemented for S3")
+	sh, err := s.s3Client(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	uploader := s3manager.NewUploaderWithClient(sh)
+
+	return &Writer{
+		bucket:   s.Bucket,
+		key:      path,
+		ctx:      ctx,
+		uploader: uploader,
+	}, nil
 }
 
 // Delete implements FS.
 func (s *S3) Delete(ctx context.Context, path string) error {
-	return fmt.Errorf("Delete not implemented for S3")
+	sh, err := s.s3Client(ctx)
+	if err != nil {
+		return err
+	}
+
+	if _, err := sh.DeleteObjectWithContext(ctx, &s3.DeleteObjectInput{
+		Bucket: aws.String(s.Bucket),
+		Key:    aws.String(path),
+	}); err != nil {
+		return fmt.Errorf("s3: unable to delete object: %v", err)
+	}
+
+	return nil
 }
 
 // Walk implements FS.
 func (s *S3) Walk(ctx context.Context, path string, fn WalkFn) error {
-	return fmt.Errorf("Walk not implemented for S3")
+	sh, err := s.s3Client(ctx)
+	if err != nil {
+		return err
+	}
+
+	errCh := make(chan error, 1)
+
+	err = sh.ListObjectsPagesWithContext(ctx, &s3.ListObjectsInput{
+		Bucket: aws.String(s.Bucket),
+		Prefix: aws.String(path),
+	}, func(page *s3.ListObjectsOutput, last bool) bool {
+		for _, obj := range page.Contents {
+			if err := fn(*obj.Key); err != nil {
+				errCh <- err
+				return false
+			}
+		}
+		return last
+	})
+
+	if err != nil {
+		return fmt.Errorf("s3: unable to walk: %v", err)
+	}
+
+	close(errCh)
+	return <-errCh
+}
+
+func (s *S3) s3Client(ctx context.Context) (*s3.S3, error) {
+	sess, err := session.NewSession()
+	if err != nil {
+		return nil, fmt.Errorf("s3: unable to create session: %v", err)
+	}
+
+	// https://docs.aws.amazon.com/sdk-for-go/api/service/s3/s3manager/#GetBucketRegion
+	region := aws.StringValue(sess.Config.Region)
+	if len(region) == 0 {
+		region, err = s3manager.GetBucketRegion(context.Background(), sess, s.Bucket, endpoints.UsEast1RegionID)
```

Comment:
`context.Background()` -> `ctx`

Have `endpoints.UsEast1RegionID` in a non-exported const outside the method so that it's more obviously a "static" parameter.